mne/decoding/gsoc.py

```diff
@@ -0,0 +1,249 @@
+
+    """Implementation of the Xdawn Algorithm.
+
+    Xdawn is a spatial filtering method designed to improve the signal
+    to signal + noise ratio (SSNR) of the ERP responses. Xdawn was originally
+    designed for P300 evoked potential by enhancing the target response with
+    respect to the non-target response. This implementation is a light version
+    of the original version. This does not handle overlap cases however it is
+    helpful while pipelining.
+
+    Parameters
+    ----------
+    n_chan : int
+         Integer indicating the number of channels.
+    n_components : int (default 2)
+        The number of components to decompose M/EEG signals.
+    signal_cov : None | Covariance | ndarray, shape (n_channels, n_channels)
+        (default None). The signal covariance used for whitening of the data.
+        if None, the covariance is estimated from the epochs signal.
+    reg : float | str | None (default None)
+        if not None, allow regularization for covariance estimation
+        if float, shrinkage covariance is used (0 <= shrinkage <= 1).
+        if str, optimal shrinkage using Ledoit-Wolf Shrinkage ('ledoit_wolf')
+        or Oracle Approximating Shrinkage ('oas').
+    correct_overlap : bool (default False)
+        Apply correction for overlaped ERP for the estimation of evokeds
+        responses. if 'auto', the overlapp correction is chosen in function
+        of the events in epochs.events.
+
+    Attributes
+    ----------
+    filters_ : dict of ndarray
+        If fit, the Xdawn components used to decompose the data for each event
+        type, else empty.
+    patterns_ : dict of ndarray
+        If fit, the Xdawn patterns used to restore M/EEG signals for each event
+        type, else empty.
+    event_id_mean_ : dict of np.ndarray
+        If fit, the mean of each event type.
+
+    See Also
+    --------
+    ICA
+    CSP
+
+    References
+    ----------
+    [1] Rivet, B., Souloumiac, A., Attina, V., & Gibert, G. (2009). xDAWN
+    algorithm to enhance evoked potentials: application to brain-computer
+    interface. Biomedical Engineering, IEEE Transactions on, 56(8), 2035-2043.
+
+    [2] Rivet, B., Cecotti, H., Souloumiac, A., Maby, E., & Mattout, J. (2011,
+    August). Theoretical analysis of xDAWN algorithm: application to an
+    efficient sensor selection in a P300 BCI. In Signal Processing Conference,
+    2011 19th European (pp. 1382-1386). IEEE.
+    """
+
+    def __init__(self, n_chan, n_components=2, signal_cov=None,
+                 reg=None, correct_overlap=False):
+        """init xdawn."""
+        self.n_chan = n_chan
+        self.n_components = n_components
+        self.signal_cov = signal_cov
+        self.reg = reg
+        if correct_overlap:
+            raise DeprecationWarning("correct_overlap feature is deprecated "
+                                     "temporarily.")
+        self.filters_ = dict()
+        self.patterns_ = dict()
+        self.event_id_mean_ = dict()
```

Comment:
Isn't `self.filters_` used in `transform` and `self.patterns_` in `apply` method?